File: src/main/java/com/hp/jipp/encoding/ResolutionType.java
package com.hp.jipp.encoding;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

public class ResolutionType extends AttributeType<Resolution> {
    private static final String TYPE_NAME = "Resolution";

    private static final int INT_LENGTH = 4;
    private static final int BYTE_LENGTH = 1;

    static final Attribute.Encoder<Resolution> ENCODER = new Attribute.Encoder<Resolution>(TYPE_NAME) {
        @Override
        public Resolution readValue(DataInputStream in, Tag valueTag) throws IOException {
            Attribute.expectLength(in, INT_LENGTH + INT_LENGTH + BYTE_LENGTH);
            return Resolution.of(in.readInt(), in.readInt(),
                    Resolution.Unit.ENCODER.get(in.readByte()));
        }

        @Override
        public void writeValue(DataOutputStream out, Resolution value) throws IOException {
            out.writeShort(INT_LENGTH + INT_LENGTH + BYTE_LENGTH);
            out.writeInt(value.getCrossFeedResolution());
            out.writeInt(value.getFeedResolution());
            out.writeByte((byte) value.getUnit().getCode());
        }

        @Override
        public boolean valid(Tag valueTag) {
            return Tag.Resolution.equals(valueTag);
        }
    };

    public ResolutionType(Tag tag, String name) {
        super(ENCODER, tag, name);
    }
}

File: src/main/java/com/hp/jipp/encoding/ResolutionType.kt
// Copyright 2017 HP Development Company, L.P.
// SPDX-License-Identifier: MIT

package com.hp.jipp.encoding

import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException

/** Attribute type for encoding of a [Resolution] */
class ResolutionType(tag: Tag, override val name: String) : AttributeType<Resolution>(Encoder, tag) {

    companion object Encoder : SimpleEncoder<Resolution>("resolution") {

        private const val INT_LENGTH = 4
        private const val BYTE_LENGTH = 1

        @Throws(IOException::class)
        override fun readValue(input: DataInputStream, valueTag: Tag): Resolution {
            input.takeLength(INT_LENGTH + INT_LENGTH + BYTE_LENGTH)
            return Resolution(input.readInt(), input.readInt(),
                    ResolutionUnit.ENCODER[input.readByte().toInt()])
        }

        @Throws(IOException::class)
        override fun writeValue(out: DataOutputStream, value: Resolution) {
            out.writeShort(INT_LENGTH + INT_LENGTH + BYTE_LENGTH)
            out.writeInt(value.crossFeedResolution)
            out.writeInt(value.feedResolution)
            out.writeByte(value.unit.code.toByte().toInt())
        }

        override fun valid(valueTag: Tag) = Tag.resolution == valueTag
    }
}
